unicon_backend/evaluator/tasks/programming/steps.py

```diff
@@ -0,0 +1,246 @@
+import abc
+from collections.abc import Iterable
+from enum import Enum
+from typing import ClassVar, Self, Union
+
+from pydantic import model_validator
+
+from unicon_backend.evaluator.tasks.programming.artifact import File, PrimitiveData
+from unicon_backend.lib.common import CustomBaseModel
+from unicon_backend.lib.graph import Graph, GraphEdge, GraphNode, NodeSocket
+
+type SocketName = str
+type ProgramVariable = str
+type ProgramFragment = str
+type AssembledProgram = str
+
+# A program can be made up of sub programs, especially with subgraphs
+Program = list[Union["Program", ProgramFragment]]
+
+
+class StepType(str, Enum):
+    PY_RUN_FUNCTION = "PY_RUN_FUNCTION_STEP"
+
+    # I/O Operations
+    INPUT = "INPUT_STEP"
+    OUTPUT = "OUTPUT_STEP"
+
+    # Control Flow Operations
+    LOOP_STEP = "LOOP_STEP"
+    BREAKING_CONDITION = "BREAKING_CONDITION_STEP"
+
+    # Comparison Operations
+    STRING_MATCH = "STRING_MATCH_STEP"
+
+
+class StepSocket(NodeSocket):
+    # The data that the socket holds
+    data: PrimitiveData | File | None = None
```

Comment:
i think i am a bit confused by StepSocket vs NodeSocket and this data variable

1. when do we use NodeSockets?
2. is data for typechecking? i thought it was, but PrimitiveData seems to be an actual value instead of specifying a type - if it's not for typechecking, how is the value being "passed around", given that we are compiling a program to run instead of running it node by node?